README.md
[![Build Status](https://travis-ci.com/profjsb/deepCR.svg?token=baKtC9yCzzwzzqM9ihAX&branch=master)](https://travis-ci.com/profjsb/deepCR) [![codecov](https://codecov.io/gh/profjsb/deepCR/branch/master/graph/badge.svg?token=SIwJFmKJqr)](https://codecov.io/gh/profjsb/deepCR)
[![Documentation Status](https://readthedocs.org/projects/deepcr/badge/?version=latest)](https://deepcr.readthedocs.io/en/latest/?badge=latest) [![DOI](https://joss.theoj.org/papers/10.21105/joss.01651/status.svg)](https://doi.org/10.21105/joss.01651) [![arXiv](https://img.shields.io/badge/astro--ph-1907.09500-blue)](https://arxiv.org/abs/1907.09500) 

## deepCR: Deep Learning Based Cosmic Ray Removal for Astronomical Images

Identify and remove cosmic rays from astronomical images using trained convolutional neural networks.

This package is implements the method described in the paper:
  > [deepCR: Cosmic Ray Rejection with Deep Learning](https://arxiv.org/abs/1907.09500)\
  > Keming Zhang & Joshua Bloom\
  > _arXiv:1907.09500; ApJ in press_
  
If you use this package, please cite the paper above and consider including a
link to this repository.

[Documentation and tutorials](https://deepcr.readthedocs.io)

[Currently available models](https://deepcr.readthedocs.io/en/latest/model_zoo.html)


<img src="https://raw.githubusercontent.com/profjsb/deepCR/master/imgs/postage-sm.jpg" wdith="90%">

### New for v0.2.0

[DECam](https://deepcr.readthedocs.io/en/latest/model_zoo.html#decam) deepCR model now available!

```python
from deepCR import deepCR
decam_model = deepCR(mask='decam', device='CPU')
```
Note 1: Model is trained on g-band images but is expected to work on 
other filters as well. We are working on benchmarking on different filters 
but before that's done please proceed with caution working with other filters.

Note 2: Inpainting model is TBA for DECam.

### Installation

```bash
pip install deepCR
```

Or you can install from source:

```bash
git clone https://github.com/profjsb/deepCR.git
cd deepCR/
pip install .
```

### Quick Start

Quick download of a HST ACS/WFC image

```bash
wget -O jdba2sooq_flc.fits https://mast.stsci.edu/api/v0.1/Download/file?uri=mast:HST/product/jdba2sooq_flc.fits
```

With Python >=3.5:

For smaller sized images
```python
from deepCR import deepCR
from astropy.io import fits
image = fits.getdata("jdba2sooq_flc.fits")[:512,:512]

# create an instance of deepCR with specified model configuration
mdl = deepCR(mask="ACS-WFC-F606W-2-32",
	     inpaint="ACS-WFC-F606W-2-32",
             device="CPU")

# apply to input image
mask, cleaned_image = mdl.clean(image, threshold = 0.5)
# best threshold is highest value that generate mask covering full extent of CR
# choose threshold by visualizing outputs.
# note that deepCR-inpaint would overestimate if mask does not fully cover CR.

# if you only need CR mask you may skip image inpainting for shorter runtime
mask = mdl.clean(image, threshold = 0.5, inpaint=False)

# if you want probabilistic cosmic ray mask instead of binary mask
prob_mask = mdl.clean(image, binary=False)
```

For WFC full size images (4k * 2k), you should specify **segment = True** to tell deepCR to segment the input image into 256*256 patches, and process one patch at a time.
Otherwise this would take up > 10gb memory. We recommended you use segment = True for images larger than 1k * 1k on CPU. GPU memory limits may be more strict.
```python
image = fits.getdata("jdba2sooq_flc.fits")
mask, cleaned_image = mdl.clean(image, threshold = 0.5, segment = True)
```

(CPU only) In place of **segment = True**, you can also specify **parallel = True** and invoke the multi-threaded version of segment mode. This will speed things up. You don't have to specify segment = True again.
```python
image = fits.getdata("jdba2sooq_flc.fits")
mask, cleaned_image = mdl.clean(image, threshold = 0.5, parallel = True, n_jobs=-1)
```
**n_jobs=-1** makes use of all your CPU cores.

Note that this won't speed things up if you're using GPU!

### Contributing

We are very interested in getting bug fixes, new functionality, and new trained models from the community (especially for ground-based imaging and spectroscopy). Please fork this repo and issue a PR with your changes. It will be especially helpful if you add some tests for your changes.


CHANGES.txt
v0.1.0, July 16, 2019 -- initial release

v0.1.1, v0.1.2, July 21, 2019 -- documentation for PyPi

v0.1.3, v0.1.4, July 22, 2019 -- arXiv release version

v0.1.5, August 13, 2019 -- added new module: deepCR.train

v0.2.0, September 28, 2019 -- added DECam model and additional feature for training API


deepCR/train.py
"""main module to instantiate deepCR models and use them
"""

import torch
from tqdm import tqdm_notebook as tqdm


from deepCR.unet import WrappedModel, UNet2Sigmoid
from learned_models import mask_dict, inpaint_dict, default_model_path

import torch.nn.functional as F
import matplotlib.pyplot as plt
import torch.nn as nn
import torch.tensor as tensor
import torch.optim as optim
from torch.autograd import Variable
from torch.utils.data import DataLoader
import numpy as np
import datetime
from torch.optim.lr_scheduler import MultiStepLR, ReduceLROnPlateau
from deepCR.util import maskMetric
from deepCR.dataset import dataset

__all__ = ('train')

class train():

    def __init__(self, image, mask, use=None, name='model', hidden=32, gpu=False, epoch=50, batch_size=16, lr=0.005, verbose=True, save_after=0, every=10):

        if use is None:
            use = np.zeros_like(image)
        data_train = dataset(image, mask, use, part='train')
        data_val = dataset(image, mask, use, part='val')
        self.TrainLoader = DataLoader(data_train, batch_size=batch_size, shuffle=False, num_workers=1)
        self.ValLoader = DataLoader(data_val, batch_size=batch_size, shuffle=False, num_workers=1)
        self.name = name
        self.verbose = verbose

        if gpu:
            self.dtype = torch.cuda.FloatTensor
            self.dint = torch.cuda.ByteTensor
            self.network = nn.DataParallel(UNet2Sigmoid(1,1,hidden))
            self.network.type(self.dtype)
        else:
            self.dtype = torch.FloatTensor
            self.dint = torch.ByteTensor
            self.network = WrappedModel(UNet2Sigmoid(1,1,hidden))
            self.network.type(self.dtype)

        self.optimizer = optim.Adam(self.network.parameters(), lr=lr)
        self.lr_scheduler = ReduceLROnPlateau(self.optimizer, factor=0.1, patience=4, cooldown=2,
                                                          min_lr=lr * 1e-4, verbose=True, threshold=0.01)

        self.BCELoss = nn.BCELoss()
        self.lossMask_train = []
        self.lossMask_val = []
        self.itr = 0
        self.epoch_mask = 0
        self.save_after = save_after
        self.n_epochs = epoch
        self.every = every

    def set_input(self, img0, mask, use):
        self.img0 = Variable(img0.type(self.dtype)).view(-1, 1, 256, 256)
        self.mask = Variable(mask.type(self.dtype)).view(-1, 1, 256, 256)
        self.use = Variable(use.type(self.dtype)).view(-1, 1, 256, 256)

    def validate_mask(self):
        lmask = 0; count = 0
        metric = np.zeros(4)
        for i, dat in enumerate(self.ValLoader):
            n = dat[0].shape[0]
            count += n
            self.set_input(*dat)
            self.pdt_mask = self.network(self.img0)
            loss = self.backward_network()
            lmask += float(loss.detach()) * n
            metric += maskMetric(self.pdt_mask.reshape(-1, 256, 256).detach().cpu().numpy() > 0.1, dat[1].numpy())
        lmask /= count
        TP, TN, FP, FN = metric[0], metric[1], metric[2], metric[3]
        TPR = TP / (TP + FN)
        FPR = FP / (FP + TN)
        print('TPR=%.3f   FPR=%.3f' % (TPR * 100, FPR * 100))
        return (lmask)
    
    def train(self):
        self.network.train()
        for epoch in tqdm(range(self.n_epochs)):
            for t, dat in enumerate(self.TrainLoader):
                self.optimize_network(dat)
            self.epoch_mask += 1

            if self.epoch_mask % self.every==0:
                plt.figure(figsize=(10, 30))
                plt.subplot(131)
                plt.imshow(np.log(self.img0[0, 0].detach().cpu().numpy()), cmap='gray')
                plt.subplot(132)
                plt.imshow(self.pdt_mask[0, 0].detach().cpu().numpy()>0.5, cmap='gray')
                plt.subplot(133)
                plt.imshow(self.mask[0, 0].detach().cpu().numpy(), cmap='gray')
                plt.show()

            print('epoch = %d' % (self.epoch_mask))
            valLossMask = self.validate_mask()
            self.lossMask_val.append(valLossMask)
            if (np.array(self.lossMask_val)[-1] == np.array(
                    self.lossMask_val).min() and self.epoch_mask > self.save_after):
                self.save_mask()
            self.lr_scheduler.step(self.lossMask_val[-1])
            print('loss = %.4f' % (self.lossMask_val[-1]))
            self.n_epochs -= 1

    def optimize_network(self, dat):
        self.set_input(*dat)
        self.itr += self.img0.shape[0]
        self.pdt_mask = self.network(self.img0)
        self.optimizer.zero_grad()
        loss = self.backward_network()
        loss.backward()
        self.optimizer.step()
        self.lossMask_train.append(float(loss.detach()))

    def backward_network(self):
        loss = self.BCELoss(self.pdt_mask * (1 - self.use), self.mask * (1 - self.use))
        return (loss)

    def save_mask(self):
        time = datetime.datetime.now()
        time = str(time)[:10]
        filename = '%s_%s_epoch%d' % (time, self.name, self.epoch_mask)
        print(filename)
        torch.save(self.network.state_dict(), filename + '.pth')
        lossTrain = np.array(self.lossMask_train)
        lossVal = np.array(self.lossMask_val)
        np.save(filename + '_mask_train.npy', lossTrain)
        np.save(filename + '_mask_val.npy', lossVal)

    def load_mask(self, filename):
        # self.load_model_mask(filename)
        self.network.load_state_dict(torch.load(filename + '.pth'))
        # self.load_loss_mask(filename)
        self.lossMask_train = list(np.load(filename + '_mask_train.npy'))
        self.lossMask_val = list(np.load(filename + '_mask_val.npy'))
        loc = filename.find('epoch') + 5
        self.epoch_mask = int(filename[loc:])